In this notebook,
we trying to solve below three problems.

<b> Q.2.1 </b> Data extraction 
○ Please extract the following fields and store the data as .csv
i. fields required:
1. Restaurant id,
2. Restaurant name,
3. Country name,
4. City name
5. User rating votes,
6. User aggregate_rating (in float),
7. Cuisines

</br> <b> Q.2.2 </b> Extract list of restaurants that have past event within the month of April 2017 and
store the data as .csv
i. Fields required:
1. Event id
2. Restaurant id
3. Restaurant name
4. Photo url
5. Event title
6. Event start date
7. Event end date
ii. If there are fields with empty value, please populate with “NA”

</br><b> Q.4. </b> Optional ETL challenge:
○ From the dataset , use data analytics to determine the threshold for
the different “rating_text” based on “aggregate_rating”

##### Import required python libraries

In [429]:
import pandas as pd
import json

### Reading Restaurant_data.json file and loading into FileDf Dataframe.

In [430]:
FileDf = pd.read_json("https://raw.githubusercontent.com/ashraf356/cc4braininterview/main/restaurant_data.json")
FileDf.head()

results_found                                        restaurants  \
0        17151.0  [{'restaurant': {'has_online_delivery': 1, 'ph...   
1         4748.0  [{'restaurant': {'has_online_delivery': 0, 'ph...   
2        13786.0  [{'restaurant': {'has_online_delivery': 0, 'ph...   
3        10224.0  [{'restaurant': {'has_online_delivery': 0, 'ph...   
4         7039.0  [{'restaurant': {'has_online_delivery': 0, 'ph...   

   results_shown  results_start message  code status  
0           20.0            1.0     NaN   NaN    NaN  
1           20.0            1.0     NaN   NaN    NaN  
2           20.0            1.0     NaN   NaN    NaN  
3           20.0            1.0     NaN   NaN    NaN  
4           20.0            1.0     NaN   NaN    NaN

#### Number of rows and columns in FileDf is 479, 7 respectively

In [431]:
FileDf.shape

(479, 7)

###### We can observe that in FileDf , column 'restaurants' has list of elements. To  transform each element of a list-like to a row, we use here explode( ) function.

In [432]:
FileDf = FileDf.explode('restaurants')
FileDf.shape

(1600, 7)

##### After explode() number of rows increases to 1600, columns are still 7.

In [417]:
FileDf.head()

results_found                                        restaurants  \
0        17151.0  {'restaurant': {'has_online_delivery': 1, 'pho...   
0        17151.0  {'restaurant': {'has_online_delivery': 0, 'pho...   
0        17151.0  {'restaurant': {'has_online_delivery': 1, 'pho...   
0        17151.0  {'restaurant': {'has_online_delivery': 0, 'pho...   
0        17151.0  {'restaurant': {'has_online_delivery': 0, 'pho...   

   results_shown  results_start message  code status  
0           20.0            1.0     NaN   NaN    NaN  
0           20.0            1.0     NaN   NaN    NaN  
0           20.0            1.0     NaN   NaN    NaN  
0           20.0            1.0     NaN   NaN    NaN  
0           20.0            1.0     NaN   NaN    NaN

##### Now creat new Dataframe 'RestaurantsDf' by converting FileDf to json then load and then normalize to get different columns from list like rows. 

In [433]:
RestaurantsDf = pd.json_normalize(json.loads(FileDf.to_json(orient = "records")))
RestaurantsDf.shape

(1600, 45)

#####  'RestaurantsDf' dataframe has 1600 rows and 45 columns

In [420]:
RestaurantsDf.head()

results_found  results_shown  results_start message  code status  \
0        17151.0           20.0            1.0    None   NaN   None   
1        17151.0           20.0            1.0    None   NaN   None   
2        17151.0           20.0            1.0    None   NaN   None   
3        17151.0           20.0            1.0    None   NaN   None   
4        17151.0           20.0            1.0    None   NaN   None   

   restaurants.restaurant.has_online_delivery  \
0                                         1.0   
1                                         0.0   
2                                         1.0   
3                                         0.0   
4                                         0.0   

                   restaurants.restaurant.photos_url  \
0  https://www.zomato.com/HauzKhasSocial/photos?u...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
2  https://www.zomato.com/ncr/the-hudson-cafe-del...   
3  https://www.zomato.com/ncr/summer-house-cafe-h...   
4  https://www.zomato.com/ncr/38-barracks-connaug...   

                          restaurants.restaurant.url  \
0  https://www.zomato.com/HauzKhasSocial?utm_sour...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
2  https://www.zomato.com/ncr/the-hudson-cafe-del...   
3  https://www.zomato.com/ncr/summer-house-cafe-h...   
4  https://www.zomato.com/ncr/38-barracks-connaug...   

   restaurants.restaurant.price_range  ...  \
0                                 3.0  ...   
1                                 3.0  ...   
2                                 2.0  ...   
3                                 3.0  ...   
4                                 3.0  ...   

               restaurants.restaurant.featured_image  \
0  https://b.zmtcdn.com/data/pictures/2/308322/cf...   
1  https://b.zmtcdn.com/data/pictures/7/18037817/...   
2  https://b.zmtcdn.com/data/pictures/5/312345/03...   
3  https://b.zmtcdn.com/data/pictures/0/307490/e0...   
4  https://b.zmtcdn.com/data/pictures/7/18241537/...   

                restaurants.restaurant.zomato_events  \
0  [{'event': {'display_date': '05 April - 21 May...   
1  [{'event': {'display_date': '16 April - 17 Apr...   
2                                                NaN   
3  [{'event': {'display_date': '16 April - 17 Apr...   
4                                                NaN   

  restaurants.restaurant.currency restaurants.restaurant.id  \
0                             Rs.                    308322   
1                             Rs.                  18037817   
2                             Rs.                    312345   
3                             Rs.                    307490   
4                             Rs.                  18241537   

                        restaurants.restaurant.thumb  \
0  https://b.zmtcdn.com/data/pictures/2/308322/cf...   
1  https://b.zmtcdn.com/data/pictures/7/18037817/...   
2  https://b.zmtcdn.com/data/pictures/5/312345/03...   
3  https://b.zmtcdn.com/data/pictures/0/307490/e0...   
4  https://b.zmtcdn.com/data/pictures/7/18241537/...   

   restaurants.restaurant.establishment_types  \
0                                          []   
1                                          []   
2                                          []   
3                                          []   
4                                          []   

                   restaurants.restaurant.events_url  \
0  https://www.zomato.com/HauzKhasSocial/events#t...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
2  https://www.zomato.com/ncr/the-hudson-cafe-del...   
3  https://www.zomato.com/ncr/summer-house-cafe-h...   
4  https://www.zomato.com/ncr/38-barracks-connaug...   

  restaurants.restaurant.order_deeplink  restaurants.restaurant.order_url  \
0                                   NaN                               NaN   
1                                   NaN                               NaN   
2                                   NaN                               NaN   


##### similar as above,  'RestaurantsDf' dataframe's column 'restaurants.restaurant.zomato_events'  has list of elements.Repeating above process to get normalize form.

In [434]:
RestaurantsDf = RestaurantsDf.explode('restaurants.restaurant.zomato_events')
RestaurantsDf.shape

(1651, 45)

In [435]:
RestaurantsDf.head()

results_found  results_shown  results_start message  code status  \
0        17151.0           20.0            1.0    None   NaN   None   
1        17151.0           20.0            1.0    None   NaN   None   
1        17151.0           20.0            1.0    None   NaN   None   
2        17151.0           20.0            1.0    None   NaN   None   
3        17151.0           20.0            1.0    None   NaN   None   

   restaurants.restaurant.has_online_delivery  \
0                                         1.0   
1                                         0.0   
1                                         0.0   
2                                         1.0   
3                                         0.0   

                   restaurants.restaurant.photos_url  \
0  https://www.zomato.com/HauzKhasSocial/photos?u...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
2  https://www.zomato.com/ncr/the-hudson-cafe-del...   
3  https://www.zomato.com/ncr/summer-house-cafe-h...   

                          restaurants.restaurant.url  \
0  https://www.zomato.com/HauzKhasSocial?utm_sour...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
2  https://www.zomato.com/ncr/the-hudson-cafe-del...   
3  https://www.zomato.com/ncr/summer-house-cafe-h...   

   restaurants.restaurant.price_range  ...  \
0                                 3.0  ...   
1                                 3.0  ...   
1                                 3.0  ...   
2                                 2.0  ...   
3                                 3.0  ...   

               restaurants.restaurant.featured_image  \
0  https://b.zmtcdn.com/data/pictures/2/308322/cf...   
1  https://b.zmtcdn.com/data/pictures/7/18037817/...   
1  https://b.zmtcdn.com/data/pictures/7/18037817/...   
2  https://b.zmtcdn.com/data/pictures/5/312345/03...   
3  https://b.zmtcdn.com/data/pictures/0/307490/e0...   

                restaurants.restaurant.zomato_events  \
0  {'event': {'display_date': '05 April - 21 May'...   
1  {'event': {'display_date': '16 April - 17 Apri...   
1  {'event': {'display_date': '19 April - 20 Apri...   
2                                                NaN   
3  {'event': {'display_date': '16 April - 17 Apri...   

  restaurants.restaurant.currency restaurants.restaurant.id  \
0                             Rs.                    308322   
1                             Rs.                  18037817   
1                             Rs.                  18037817   
2                             Rs.                    312345   
3                             Rs.                    307490   

                        restaurants.restaurant.thumb  \
0  https://b.zmtcdn.com/data/pictures/2/308322/cf...   
1  https://b.zmtcdn.com/data/pictures/7/18037817/...   
1  https://b.zmtcdn.com/data/pictures/7/18037817/...   
2  https://b.zmtcdn.com/data/pictures/5/312345/03...   
3  https://b.zmtcdn.com/data/pictures/0/307490/e0...   

   restaurants.restaurant.establishment_types  \
0                                          []   
1                                          []   
1                                          []   
2                                          []   
3                                          []   

                   restaurants.restaurant.events_url  \
0  https://www.zomato.com/HauzKhasSocial/events#t...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
2  https://www.zomato.com/ncr/the-hudson-cafe-del...   
3  https://www.zomato.com/ncr/summer-house-cafe-h...   

  restaurants.restaurant.order_deeplink  restaurants.restaurant.order_url  \
0                                   NaN                               NaN   
1                                   NaN                               NaN   
1                                   NaN                               NaN   


In [436]:
EventDf = pd.json_normalize(json.loads(RestaurantsDf.to_json(orient = "records")))
EventDf.shape

(1651, 67)

In [437]:
EventDf.head()

results_found  results_shown  results_start message  code status  \
0        17151.0           20.0            1.0    None   NaN   None   
1        17151.0           20.0            1.0    None   NaN   None   
2        17151.0           20.0            1.0    None   NaN   None   
3        17151.0           20.0            1.0    None   NaN   None   
4        17151.0           20.0            1.0    None   NaN   None   

   restaurants.restaurant.has_online_delivery  \
0                                         1.0   
1                                         0.0   
2                                         0.0   
3                                         1.0   
4                                         0.0   

                   restaurants.restaurant.photos_url  \
0  https://www.zomato.com/HauzKhasSocial/photos?u...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
2  https://www.zomato.com/ncr/qubitos-the-terrace...   
3  https://www.zomato.com/ncr/the-hudson-cafe-del...   
4  https://www.zomato.com/ncr/summer-house-cafe-h...   

                          restaurants.restaurant.url  \
0  https://www.zomato.com/HauzKhasSocial?utm_sour...   
1  https://www.zomato.com/ncr/qubitos-the-terrace...   
2  https://www.zomato.com/ncr/qubitos-the-terrace...   
3  https://www.zomato.com/ncr/the-hudson-cafe-del...   
4  https://www.zomato.com/ncr/summer-house-cafe-h...   

   restaurants.restaurant.price_range  ...  \
0                                 3.0  ...   
1                                 3.0  ...   
2                                 3.0  ...   
3                                 2.0  ...   
4                                 3.0  ...   

  restaurants.restaurant.zomato_events.event.is_end_time_set  \
0                                                1.0           
1                                                1.0           
2                                                1.0           
3                                                NaN           
4                                                1.0           

  restaurants.restaurant.zomato_events.event.event_id  \
0                                           118139.0    
1                                           119563.0    
2                                           120517.0    
3                                                NaN    
4                                           100228.0    

  restaurants.restaurant.zomato_events.event.end_date  \
0                                         2017-05-21    
1                                         2017-04-17    
2                                         2017-04-20    
3                                                NaN    
4                                         2017-04-17    

  restaurants.restaurant.zomato_events.event.event_category  \
0                                                0.0          
1                                                0.0          
2                                                0.0          
3                                                NaN          
4                                                1.0          

  restaurants.restaurant.zomato_events.event.friendly_end_date  \
0                                             21 May             
1                                           17 April             
2                                           20 April             
3                                                NaN             
4                                           17 April             

   restaurants.restaurant.zomato_events.event.is_active  \
0                                                1.0      
1                                                1.0      
2                                                1.0      
3                                                NaN      
4                                                1.0      

  restaurants.restaurant.zomato_events.event.start_time  \
0                                           10:00:00      
1         

In [438]:
EventDf = EventDf.explode('restaurants.restaurant.zomato_events.event.photos')
EventDf.shape

(1719, 67)

In [439]:
PhotoDf = pd.json_normalize(json.loads(EventDf.to_json(orient = "records")))
PhotoDf.shape

(1719, 74)

In [440]:
PhotoDf.head()

results_found  results_shown  results_start message  code status  \
0        17151.0           20.0            1.0    None   NaN   None   
1        17151.0           20.0            1.0    None   NaN   None   
2        17151.0           20.0            1.0    None   NaN   None   
3        17151.0           20.0            1.0    None   NaN   None   
4        17151.0           20.0            1.0    None   NaN   None   

   restaurants.restaurant.has_online_delivery  \
0                                         1.0   
1                                         1.0   
2                                         0.0   
3                                         0.0   
4                                         1.0   

                   restaurants.restaurant.photos_url  \
0  https://www.zomato.com/HauzKhasSocial/photos?u...   
1  https://www.zomato.com/HauzKhasSocial/photos?u...   
2  https://www.zomato.com/ncr/qubitos-the-terrace...   
3  https://www.zomato.com/ncr/qubitos-the-terrace...   
4  https://www.zomato.com/ncr/the-hudson-cafe-del...   

                          restaurants.restaurant.url  \
0  https://www.zomato.com/HauzKhasSocial?utm_sour...   
1  https://www.zomato.com/HauzKhasSocial?utm_sour...   
2  https://www.zomato.com/ncr/qubitos-the-terrace...   
3  https://www.zomato.com/ncr/qubitos-the-terrace...   
4  https://www.zomato.com/ncr/the-hudson-cafe-del...   

   restaurants.restaurant.price_range  ...  \
0                                 3.0  ...   
1                                 3.0  ...   
2                                 3.0  ...   
3                                 3.0  ...   
4                                 2.0  ...   

  restaurants.restaurant.zomato_events.event.event_category_name  \
0                                                                  
1                                                                  
2                                                                  
3                                                                  
4                                               None               

  restaurants.restaurant.zomato_events  \
0                                 None   
1                                 None   
2                                 None   
3                                 None   
4                                 None   

  restaurants.restaurant.zomato_events.event.photos.photo.photo_id  \
0                                           185854.0                 
1                                           185884.0                 
2                                           187914.0                 
3                                                NaN                 
4                                                NaN                 

  restaurants.restaurant.zomato_events.event.photos.photo.order  \
0                                                0.0              
1                                                1.0              
2                                                0.0              
3                                                NaN              
4                                                NaN              

  restaurants.restaurant.zomato_events.event.photos.photo.type  \
0                                             NORMAL             
1                                             NORMAL             
2                                             NORMAL             
3                                                NaN             
4                                                NaN             

   restaurants.restaurant.zomato_events.event.photos.photo.url  \
0  https://b.zmtcdn.com/data/zomato_events/photos...             
1  https://b.zmtcdn.com/data/zomato_events/photos...             
2  https://b.zmtcdn.com/data/zomato_events/photos...             
3                                                NaN             
4                                                NaN             

  restaurants.restaurant.zomato_events.event.photo

##### After normalizing full json file, we got total 1719 rows and 74 columns.


##### Now slicing required columns  and renaming them for further analysis.

In [441]:
Solution = PhotoDf[['restaurants.restaurant.id',
'restaurants.restaurant.name',
'restaurants.restaurant.location.country_id',
'restaurants.restaurant.location.city',
'restaurants.restaurant.user_rating.votes',
'restaurants.restaurant.user_rating.rating_text',
'restaurants.restaurant.user_rating.aggregate_rating',
'restaurants.restaurant.cuisines',

'restaurants.restaurant.zomato_events.event.event_id',
'restaurants.restaurant.zomato_events.event.title',
'restaurants.restaurant.zomato_events.event.start_date',
'restaurants.restaurant.zomato_events.event.end_date',
'restaurants.restaurant.zomato_events.event.photos.photo.url']]

In [442]:
cols = ['Restaurant_id', 'Restaurant_name','Country_id','City_name','User_rating_votes','User_rating_text','User_agg_rating','Cuisines','Event_id','Event_title',
        'Event_start_date','Event_end_date','Event_photo_url']

In [443]:
Solution.columns = cols
Solution.shape

(1719, 13)

### Reading Country.xlsx file

In [452]:
country =  pd.read_excel("https://github.com/ashraf356/cc4braininterview/blob/main/Country-Code.xlsx?raw=true")
country.shape

(15, 2)

In [453]:
country.columns = ['Country_id','Country_name']
country.head()

Country_id Country_name
0           1        India
1          14    Australia
2          30       Brazil
3          37       Canada
4          94    Indonesia

### Merging Reastuarant and country files based on 'country_id'
##### Note - Here 'Solution' is the filtered, normalized dataframe for 'Reastuarant' json file with required number of columns.

In [454]:
Sol_merge_country = (pd.merge(Solution, country, on='Country_id'))
Sol_merge_country.head()

Restaurant_id             Restaurant_name  Country_id  City_name  \
0        308322            Hauz Khas Social         1.0  New Delhi   
1        308322            Hauz Khas Social         1.0  New Delhi   
2      18037817  Qubitos - The Terrace Cafe         1.0  New Delhi   
3      18037817  Qubitos - The Terrace Cafe         1.0  New Delhi   
4        312345             The Hudson Cafe         1.0  New Delhi   

  User_rating_votes User_rating_text User_agg_rating  \
0              7931        Very Good             4.3   
1              7931        Very Good             4.3   
2               778        Excellent             4.5   
3               778        Excellent             4.5   
4              1537        Very Good             4.4   

                                            Cuisines  Event_id  \
0         Continental, American, Asian, North Indian  118139.0   
1         Continental, American, Asian, North Indian  118139.0   
2  Thai, European, Mexican, North Indian, Chinese...  119563.0   
3  Thai, European, Mexican, North Indian, Chinese...  120517.0   
4                Cafe, Italian, Continental, Chinese       NaN   

                                   Event_title Event_start_date  \
0                         IPL Match Screenings       2017-04-05   
1                         IPL Match Screenings       2017-04-05   
2       Live Fusion Night with Suboni Brothers       2017-04-16   
3  Raenit Singh Ft. Sarvagya The Band, Qubitos       2017-04-19   
4                                         None             None   

  Event_end_date                                    Event_photo_url  \
0     2017-05-21  https://b.zmtcdn.com/data/zomato_events/photos...   
1     2017-05-21  https://b.zmtcdn.com/data/zomato_events/photos...   
2     2017-04-17  https://b.zmtcdn.com/data/zomato_events/photos...   
3     2017-04-20                                                NaN   
4           None                                                NaN   

  Country_name  
0        India  
1        India  
2        India  
3        India  
4        India

#### Rearranging columns

In [455]:
rearangecols = Sol_merge_country.columns.tolist()
rearangecols = rearangecols[:3] + rearangecols[-1:] + rearangecols[3:-1]
Sol_merge_country = Sol_merge_country[rearangecols]


In [456]:
Sol_merge_country.head()

Restaurant_id             Restaurant_name  Country_id Country_name  \
0        308322            Hauz Khas Social         1.0        India   
1        308322            Hauz Khas Social         1.0        India   
2      18037817  Qubitos - The Terrace Cafe         1.0        India   
3      18037817  Qubitos - The Terrace Cafe         1.0        India   
4        312345             The Hudson Cafe         1.0        India   

   City_name User_rating_votes User_rating_text User_agg_rating  \
0  New Delhi              7931        Very Good             4.3   
1  New Delhi              7931        Very Good             4.3   
2  New Delhi               778        Excellent             4.5   
3  New Delhi               778        Excellent             4.5   
4  New Delhi              1537        Very Good             4.4   

                                            Cuisines  Event_id  \
0         Continental, American, Asian, North Indian  118139.0   
1         Continental, American, Asian, North Indian  118139.0   
2  Thai, European, Mexican, North Indian, Chinese...  119563.0   
3  Thai, European, Mexican, North Indian, Chinese...  120517.0   
4                Cafe, Italian, Continental, Chinese       NaN   

                                   Event_title Event_start_date  \
0                         IPL Match Screenings       2017-04-05   
1                         IPL Match Screenings       2017-04-05   
2       Live Fusion Night with Suboni Brothers       2017-04-16   
3  Raenit Singh Ft. Sarvagya The Band, Qubitos       2017-04-19   
4                                         None             None   

  Event_end_date                                    Event_photo_url  
0     2017-05-21  https://b.zmtcdn.com/data/zomato_events/photos...  
1     2017-05-21  https://b.zmtcdn.com/data/zomato_events/photos...  
2     2017-04-17  https://b.zmtcdn.com/data/zomato_events/photos...  
3     2017-04-20                                                NaN  
4           None                                                NaN

##### Changing datatype of few columns

In [457]:
Sol_merge_country['User_agg_rating'] = pd.to_numeric(Sol_merge_country['User_agg_rating'], downcast="float")
Sol_merge_country['Event_start_date'] = pd.to_datetime(Sol_merge_country['Event_start_date'])
Sol_merge_country['Event_end_date'] = pd.to_datetime(Sol_merge_country['Event_end_date'])
Sol_merge_country.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1279 entries, 0 to 1278
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Restaurant_id      1279 non-null   object        
 1   Restaurant_name    1279 non-null   object        
 2   Country_id         1279 non-null   float64       
 3   Country_name       1279 non-null   object        
 4   City_name          1279 non-null   object        
 5   User_rating_votes  1279 non-null   object        
 6   User_rating_text   1279 non-null   object        
 7   User_agg_rating    1279 non-null   float32       
 8   Cuisines           1279 non-null   object        
 9   Event_id           196 non-null    float64       
 10  Event_title        196 non-null    object        
 11  Event_start_date   196 non-null    datetime64[ns]
 12  Event_end_date     196 non-null    datetime64[ns]
 13  Event_photo_url    181 non-null    object        
dtypes: datet

In [458]:
Sol_merge_country.head()

Restaurant_id             Restaurant_name  Country_id Country_name  \
0        308322            Hauz Khas Social         1.0        India   
1        308322            Hauz Khas Social         1.0        India   
2      18037817  Qubitos - The Terrace Cafe         1.0        India   
3      18037817  Qubitos - The Terrace Cafe         1.0        India   
4        312345             The Hudson Cafe         1.0        India   

   City_name User_rating_votes User_rating_text  User_agg_rating  \
0  New Delhi              7931        Very Good              4.3   
1  New Delhi              7931        Very Good              4.3   
2  New Delhi               778        Excellent              4.5   
3  New Delhi               778        Excellent              4.5   
4  New Delhi              1537        Very Good              4.4   

                                            Cuisines  Event_id  \
0         Continental, American, Asian, North Indian  118139.0   
1         Continental, American, Asian, North Indian  118139.0   
2  Thai, European, Mexican, North Indian, Chinese...  119563.0   
3  Thai, European, Mexican, North Indian, Chinese...  120517.0   
4                Cafe, Italian, Continental, Chinese       NaN   

                                   Event_title Event_start_date  \
0                         IPL Match Screenings       2017-04-05   
1                         IPL Match Screenings       2017-04-05   
2       Live Fusion Night with Suboni Brothers       2017-04-16   
3  Raenit Singh Ft. Sarvagya The Band, Qubitos       2017-04-19   
4                                         None              NaT   

  Event_end_date                                    Event_photo_url  
0     2017-05-21  https://b.zmtcdn.com/data/zomato_events/photos...  
1     2017-05-21  https://b.zmtcdn.com/data/zomato_events/photos...  
2     2017-04-17  https://b.zmtcdn.com/data/zomato_events/photos...  
3     2017-04-20                                                NaN  
4            NaT                                                NaN

# Now working on problem

### Problem2_1 ○ Please extract the following fields and store the data as .csv
<b> i. fields required:</b>
<br/>1. Restaurant id,
<br/>2. Restaurant name,
<br/>3. Country name,
<br/>4. City name
<br/>5. User rating votes,
<br/>6. User aggregate_rating (in float),
<br/>7. Cuisines

In [460]:
Solution2_1 = Sol_merge_country[['Restaurant_id',
 'Restaurant_name',
'Country_name',
 'City_name',
 'User_rating_votes',
 'User_agg_rating',
 'Cuisines']]
Solution2_1.shape

(1279, 7)

In [461]:
Solution2_1 = Solution2_1.drop_duplicates()
Solution2_1.shape

(1160, 7)

In [462]:
Solution2_1.head()

Restaurant_id             Restaurant_name Country_name  City_name  \
0         308322            Hauz Khas Social        India  New Delhi   
2       18037817  Qubitos - The Terrace Cafe        India  New Delhi   
4         312345             The Hudson Cafe        India  New Delhi   
5         307490           Summer House Cafe        India  New Delhi   
10      18241537                 38 Barracks        India  New Delhi   

   User_rating_votes  User_agg_rating  \
0               7931              4.3   
2                778              4.5   
4               1537              4.4   
5               1823              4.1   
10               840              4.4   

                                             Cuisines  
0          Continental, American, Asian, North Indian  
2   Thai, European, Mexican, North Indian, Chinese...  
4                 Cafe, Italian, Continental, Chinese  
5                                Italian, Continental  
10             North Indian, Italian, Asian, American

##### Saving result into .csv file

In [463]:
Solution2_1.to_csv('1_Solution_Restaurants_Details.csv', index = False)

### Problem2_2 ○ Extract list of restaurants that have past event within the month of April 2017 and store the data as .csv
<b>i. Fields required: </b>
<br/> 1. Event id
<br/> 2. Restaurant id
<br/> 3. Restaurant name
<br/> 4. Photo url
<br/> 5. Event title
<br/> 6. Event start date
<br/> 7. Event end date
<br/>
<b> ii. If there are fields with empty value, please populate with “NA”

In [577]:
Solution2 = Sol_merge_country[['Event_id','Restaurant_id',
 'Restaurant_name','Event_photo_url','Event_title',
 'Event_start_date',
 'Event_end_date'
 ]]

Solution2.shape

(1279, 7)

In [578]:
Solution2.head()

Event_id Restaurant_id             Restaurant_name  \
0  118139.0        308322            Hauz Khas Social   
1  118139.0        308322            Hauz Khas Social   
2  119563.0      18037817  Qubitos - The Terrace Cafe   
3  120517.0      18037817  Qubitos - The Terrace Cafe   
4       NaN        312345             The Hudson Cafe   

                                     Event_photo_url  \
0  https://b.zmtcdn.com/data/zomato_events/photos...   
1  https://b.zmtcdn.com/data/zomato_events/photos...   
2  https://b.zmtcdn.com/data/zomato_events/photos...   
3                                                NaN   
4                                                NaN   

                                   Event_title Event_start_date Event_end_date  
0                         IPL Match Screenings       2017-04-05     2017-05-21  
1                         IPL Match Screenings       2017-04-05     2017-05-21  
2       Live Fusion Night with Suboni Brothers       2017-04-16     2017-04-17  
3  Raenit Singh Ft. Sarvagya The Band, Qubitos       2017-04-19     2017-04-20  
4                                         None              NaT            NaT

##### Filtering Dataframe for April 2017

In [579]:
Solution2_2 =Solution2[((Solution2['Event_start_date'].dt.year == 2017) & (Solution2['Event_start_date'].dt.month == 4)) & ((Solution2['Event_end_date'].dt.year == 2017)& (Solution2['Event_end_date'].dt.month == 4))]
Solution2_2.head()

Event_id Restaurant_id             Restaurant_name  \
2  119563.0      18037817  Qubitos - The Terrace Cafe   
3  120517.0      18037817  Qubitos - The Terrace Cafe   
5  100228.0        307490           Summer House Cafe   
6  100212.0        307490           Summer House Cafe   
7  107961.0        307490           Summer House Cafe   

                                     Event_photo_url  \
2  https://b.zmtcdn.com/data/zomato_events/photos...   
3                                                NaN   
5  https://b.zmtcdn.com/data/zomato_events/photos...   
6  https://b.zmtcdn.com/data/zomato_events/photos...   
7  https://b.zmtcdn.com/data/zomato_events/photos...   

                                   Event_title Event_start_date Event_end_date  
2       Live Fusion Night with Suboni Brothers       2017-04-16     2017-04-17  
3  Raenit Singh Ft. Sarvagya The Band, Qubitos       2017-04-19     2017-04-20  
5                                  SALSA NIGHT       2017-04-16     2017-04-17  
6                                Karaoke Night       2017-04-17     2017-04-18  
7                                 COMEDY NIGHT       2017-04-18     2017-04-19

In [580]:
Solution2_2.shape

(58, 7)

In [581]:
Solution2_2.isna().any()

Event_id            False
Restaurant_id       False
Restaurant_name     False
Event_photo_url      True
Event_title         False
Event_start_date    False
Event_end_date      False
dtype: bool

In [582]:
Solution2_2[Solution2_2['Event_photo_url'].isna()]

Event_id Restaurant_id             Restaurant_name Event_photo_url  \
3    120517.0      18037817  Qubitos - The Terrace Cafe             NaN   
54   117303.0      18435740              38 Degree East             NaN   
62   120501.0      18447068                  Cafe Hydro             NaN   
104  120370.0      18354483                  Farzi Cafe             NaN   
154  120010.0         72475                     Haunted             NaN   

                                     Event_title Event_start_date  \
3    Raenit Singh Ft. Sarvagya The Band, Qubitos       2017-04-19   
54                           SHOT SLURP WEEKEND        2017-04-15   
62                            Carnival pet show        2017-04-16   
104           Prateek Bhaduri Live at Farzi Cafe       2017-04-19   
154                            Omen Lunch Buffet       2017-04-15   

    Event_end_date  
3       2017-04-20  
54      2017-04-16  
62      2017-04-16  
104     2017-04-19  
154     2017-04-16

In [583]:
Solution2_2 = Solution2_2.fillna("NA")
Solution2_2.head()

Event_id Restaurant_id             Restaurant_name  \
2  119563.0      18037817  Qubitos - The Terrace Cafe   
3  120517.0      18037817  Qubitos - The Terrace Cafe   
5  100228.0        307490           Summer House Cafe   
6  100212.0        307490           Summer House Cafe   
7  107961.0        307490           Summer House Cafe   

                                     Event_photo_url  \
2  https://b.zmtcdn.com/data/zomato_events/photos...   
3                                                 NA   
5  https://b.zmtcdn.com/data/zomato_events/photos...   
6  https://b.zmtcdn.com/data/zomato_events/photos...   
7  https://b.zmtcdn.com/data/zomato_events/photos...   

                                   Event_title Event_start_date Event_end_date  
2       Live Fusion Night with Suboni Brothers       2017-04-16     2017-04-17  
3  Raenit Singh Ft. Sarvagya The Band, Qubitos       2017-04-19     2017-04-20  
5                                  SALSA NIGHT       2017-04-16     2017-04-17  
6                                Karaoke Night       2017-04-17     2017-04-18  
7                                 COMEDY NIGHT       2017-04-18     2017-04-19

##### Populating empty fields with “NA”

In [584]:
Solution2_2[Solution2_2['Event_photo_url'] == "NA"]

Event_id Restaurant_id             Restaurant_name Event_photo_url  \
3    120517.0      18037817  Qubitos - The Terrace Cafe              NA   
54   117303.0      18435740              38 Degree East              NA   
62   120501.0      18447068                  Cafe Hydro              NA   
104  120370.0      18354483                  Farzi Cafe              NA   
154  120010.0         72475                     Haunted              NA   

                                     Event_title Event_start_date  \
3    Raenit Singh Ft. Sarvagya The Band, Qubitos       2017-04-19   
54                           SHOT SLURP WEEKEND        2017-04-15   
62                            Carnival pet show        2017-04-16   
104           Prateek Bhaduri Live at Farzi Cafe       2017-04-19   
154                            Omen Lunch Buffet       2017-04-15   

    Event_end_date  
3       2017-04-20  
54      2017-04-16  
62      2017-04-16  
104     2017-04-19  
154     2017-04-16

In [586]:
Solution2_2['Event_id'] = pd.to_numeric(Solution2_2['Event_id'],downcast="integer")
Solution2_2.shape

(58, 7)

In [587]:
Solution2_2.head(2)

Event_id Restaurant_id             Restaurant_name  \
2    119563      18037817  Qubitos - The Terrace Cafe   
3    120517      18037817  Qubitos - The Terrace Cafe   

                                     Event_photo_url  \
2  https://b.zmtcdn.com/data/zomato_events/photos...   
3                                                 NA   

                                   Event_title Event_start_date Event_end_date  
2       Live Fusion Night with Suboni Brothers       2017-04-16     2017-04-17  
3  Raenit Singh Ft. Sarvagya The Band, Qubitos       2017-04-19     2017-04-20

##### Saving result into .csv file

In [588]:
Solution2_2.to_csv('1_Solution_Events_Details.csv', index = False)

### Solution4 (Optional ETL challenge:)
○ From the dataset, use data analytics to determine the threshold for the different <b>“rating_text” </b>based on <b>“aggregate_rating”

In [542]:
Solution4 = Sol_merge_country[['Restaurant_id','User_rating_text','User_agg_rating']]
Solution4 = Solution4.drop_duplicates()
Solution4 = Solution4[['User_rating_text','User_agg_rating']]
Solution4.head()

User_rating_text  User_agg_rating
0         Very Good              4.3
2         Excellent              4.5
4         Very Good              4.4
5         Very Good              4.1
10        Very Good              4.4

In [543]:
Solution4.shape

(1160, 2)

##### Considering Median of 'user_agg_ rating' for threshold

In [483]:
Solution4.groupby('User_rating_text').median()

User_agg_rating
User_rating_text                 
Average                       3.3
Excellent                     4.6
Good                          3.7
Not rated                     0.0
Poor                          2.4
Very Good                     4.2

##### Calculating 'Rating_median' and 'outlier' for every rating

In [545]:
Solution4['Rating_median'] = Solution4.groupby('User_rating_text').transform('median')
difference = (Solution4.User_agg_rating - Solution4['Rating_median']).abs()
Solution4['outlier'] = ( difference > 0.5 ) 
Solution4.head()

User_rating_text  User_agg_rating  Rating_median  outlier
0         Very Good              4.3            4.2    False
2         Excellent              4.5            4.6    False
4         Very Good              4.4            4.2    False
5         Very Good              4.1            4.2    False
10        Very Good              4.4            4.2    False

##### There are two outlier for Average rating.
As we can observe below  user_agg_rating is 2.6 and 2.5, These two ratings should belongs to "Poor", but given under "Average"

In [546]:
print( Solution4[(Solution4.outlier  == True)] )

    User_rating_text  User_agg_rating  Rating_median  outlier
251          Average              2.6            3.3     True
920          Average              2.5            3.3     True
